service/mapper/transitions_test.go

```diff
@@ -0,0 +1,1233 @@
+			FirstFunc: func(height uint64) error {
+				return mocks.DummyError
+			},
+			LastFunc: func(height uint64) error {
+				return nil
+			},
+		}
+
+		forest := &mocks.Forest{
+			ResetFunc: func(finalized flow.StateCommitment) {},
+		}
+
+		tr := &Transitions{
+			log:   testLog,
+			index: index,
+			once:  &sync.Once{},
+		}
+
+		st := &State{
+			registers: map[ledger.Path]*ledger.Payload{
+				testPath1: testPayload1,
+				testPath2: testPayload2,
+				testPath3: testPayload3,
+				testPath4: testPayload4,
+				testPath5: testPayload5,
+				testPath6: testPayload6,
+			},
+			forest: forest,
+			height: testHeight,
+			status: StatusIndexed,
+			next:   testCommit,
+		}
+
+		err := tr.ForwardHeight(st)
+
+		assert.Error(t, err)
+	})
+
+	t.Run("handles indexer error on last", func(t *testing.T) {
+		t.Parallel()
+
+		index := &mocks.Writer{
+			FirstFunc: func(height uint64) error {
+				return nil
+			},
+			LastFunc: func(height uint64) error {
+				return mocks.DummyError
+			},
+		}
+
+		forest := &mocks.Forest{
+			ResetFunc: func(finalized flow.StateCommitment) {},
+		}
+
+		tr := &Transitions{
+			log:   testLog,
+			index: index,
+			once:  &sync.Once{},
+		}
+
+		st := &State{
+			registers: map[ledger.Path]*ledger.Payload{
+				testPath1: testPayload1,
+				testPath2: testPayload2,
+				testPath3: testPayload3,
+				testPath4: testPayload4,
+				testPath5: testPayload5,
+				testPath6: testPayload6,
+			},
+			forest: forest,
+			height: testHeight,
+			status: StatusIndexed,
+			next:   testCommit,
+		}
+
+		err := tr.ForwardHeight(st)
+
+		assert.Error(t, err)
+	})
+}
+
+func TestTransitions_IndexChain(t *testing.T) {
+	testHeight := uint64(42)
+	testCommit := flow.StateCommitment{0x2a, 0x04, 0x51, 0x3c, 0xc3, 0xc9, 0xa7, 0xf2, 0xec, 0x08, 0x93, 0x56, 0x5f, 0x52, 0xc2, 0x9e, 0x19, 0xf5, 0x58, 0x88, 0x10, 0x11, 0xe1, 0x13, 0x60, 0x43, 0x9e, 0x57, 0x60, 0x18, 0xe3, 0xde}
+	testBlockID := flow.Identifier{0xd5, 0xf5, 0x0b, 0xc1, 0x7b, 0xa1, 0xea, 0xad, 0x83, 0x0c, 0x86, 0xac, 0xce, 0x64, 0x5c, 0xa6, 0xc0, 0x9f, 0xf0, 0xfe, 0xc5, 0x1c, 0x76, 0x10, 0x03, 0x1c, 0xb9, 0x99, 0xa5, 0xb0, 0xb3, 0x22}
+	testEvents := []flow.Event{
+		{TransactionID: flow.Identifier{0x1, 0x2}},
+		{TransactionID: flow.Identifier{0x3, 0x4}},
+	}
+	testHeader := &flow.Header{
+		ChainID: dps.FlowTestnet,
+		Height:  testHeight,
+	}
+	testTransaction := &flow.TransactionBody{
+		ReferenceBlockID:   flow.Identifier{},
+		Script:             nil,
+		Arguments:          nil,
+		GasLimit:           0,
+		ProposalKey:        flow.ProposalKey{},
+		Payer:              flow.Address{},
+		Authorizers:        nil,
+		PayloadSignatures:  nil,
+		EnvelopeSignatures: nil,
+	}
+	testTransactions := []*flow.TransactionBody{testTransaction}
+	indexAll := Config{
+		IndexCommit:       true,
+		IndexHeader:       true,
+		IndexTransactions: true,
+		IndexEvents:       true,
+		IndexPayloads:     true,
+	}
+
+	t.Run("nominal case", func(t *testing.T) {
```

Comment:
Would be cool to have two nominal cases, one with `index everything` and one with `index nothing`.